### PR TITLE
Inconsistency: "on_failure" vs "on-failure" 

### PR DESCRIPTION
**Pull request by yank1**
I use "on-failure" to update service , but I got "on_failure".

Related to docker/docker#24369 🐳

/cc @tiborvass @cpuguy83 @thaJeztah

🐸

Signed-off-by: Kay Yan kay.yan@daocloud.io

### CODE REVIEW

File: daemon/cluster/convert/service.go [hunk 1]
```diff
@@ -272,7 +272,7 @@ func restartPolicyToGRPC(p *types.RestartPolicy) (*swarmapi.RestartPolicy, error
 	var rp *swarmapi.RestartPolicy
 	if p != nil {
 		rp = &swarmapi.RestartPolicy{}
-		sanatizedCondition := strings.ToUpper(strings.Replace(string(p.Condition), "-", "_", -1))
+		sanatizedCondition := strings.ToUpper(p.Condition.String())
```
Comment (thaJeztah) [inline daemon/cluster/convert/service.go]: Wasn't this on purpose? From the linked PR;

> protobuf enum cannot have a dash - in their name, so I updated restartPolicyToGRPC to transform on-failure to ON_FAILURE as required for SwarmKit.

**Comment by thaJeztah** [comment]
Also, build is failing;

```
9:21:22 # github.com/docker/docker/daemon/cluster/convert
09:21:22 daemon/cluster/convert/service.go:275: p.Condition.String undefined (type swarm.RestartPolicyCondition has no field or method String)
09:22:13 Build step 'Execute shell' marked build as failure

File: daemon/cluster/convert/service.go (hunk 1 above)
Comment (thaJeztah) [inline daemon/cluster/convert/service.go]: Perhaps (if I'm correct) adding a short comment to explain this would be good though

Comment (yank1) [inline daemon/cluster/convert/service.go]: But from the Doc, it shows "on-failure".
I use "on-failure" to update service , but I got "on_failure".

Comment (thaJeztah) [inline daemon/cluster/convert/service.go]: Perhaps it should be converted back in `restartPolicyFromGRPC`; https://github.com/docker/docker/pull/27014/files#diff-69edea08b760e2a4f8eb76a11e7f2e89R256

Comment (yank1) [inline daemon/cluster/convert/service.go]: I think the  "on-failure" is right becase there are only  "on-failure" in the API doc

**Comment by aaronlehmann** [comment]
@thaJeztah is correct. The actual bug is in `restartPolicyFromGRPC`. It is not converting back to the `on-restart` notation.

`restartPolicyToGRPC` is filling in a protocol buffers message to send over GRPC. The string here will never been seen by the user, and is only a temporary value used to do a lookup in a table.

IMHO, the root cause is that we are doing string manipulation here. Frankly the current code in `restartPolicyToGRPC ` is a hack, and should be rewritten as:

```go
switch p.Condition {
case types.RestartPolicyConditionNone:
        rp.Condition = swarmapi.RestartOnNone
case types.RestartPolicyConditionOnFailure:
        rp.Condition = swarmapi.RestartOnFailure
default:
        rp.Condition = swarmapi.RestartOnAny
}
```

Then there should be an inverse version of this in `restartPolicyFromGRPC`.

**Comment by thaJeztah** [comment]
Oh, yes, @aaronlehmann's suggestion looks a lot better than the current hack

**Comment by yank1** [comment]
When I curl service, I got:



I think we should resolve it  because the 3rd party software cannot keep compatibility. If the API may change to "on-failure" in the future.

**Comment by thaJeztah** [comment]
@yank1 making the changes @aaronlehmann suggested fixes that. I just pushed a branch with those changes, see: https://github.com/docker/docker/compare/master...thaJeztah:fix-restart-policy-conversion, and put the "before" / "after" changes in the commit message; https://github.com/docker/docker/commit/35b389c5b8fc1b64c24783128a47b60bd2c6d8e0

**Comment by thaJeztah** [comment]
Thanks @yank1, I see you updated the PR. Can you squash your commits, and update the commit message? Feel free to use the message I used in my commit message if you think it's good; https://github.com/docker/docker/commit/c0a40590be0f1bec0c910b6b94c94e2e4eebd6d2

**Pull request closed without merging**